### PR TITLE
Message channels

### PR DESCRIPTION
**Pull request by kevkevinpal**
Leaving as draft right now but opening up PR for visibility

- [x]  Get endpoint to get all channels per tribe
- [x]  Get endpoint to get channel by id
- [x]  Post endpoint to create new message channel
- [x]  Delete endpoint to delete a channel
- [x]  Put endpoint to edit a channel

### CODE REVIEW

File: routes.go [hunk 1]
```diff
@@ -441,6 +444,88 @@ func deleteTribe(w http.ResponseWriter, r *http.Request) {
...
+	} else { // IF PUBKEY IN CONTEXT, MUST AUTH!
+		if pubKeyFromAuth != extractedPubkey {
+			fmt.Println("createOrEditChannel pubkeys dont match")
+			w.WriteHeader(http.StatusUnauthorized)
+			return
+		}
+	}
+
+	/*
+			existing := DB.getTribe(channel.TribeUUID)
```
Comment (kevkevinpal) [inline routes.go:500]: Do I need to generate a unique `id` for the channel like we do in the `createOrEditTribe` function for `uuid`

Comment (Evanfeenstra) [reply routes.go:540]: That createOrEditTribe function is complex for a variety of reasons, you dont need to do that for Channels.

```
func (db database) createChannel(c Channel) Channel {
	db.db.Create(&c)
	return c
}
func (db database) updateChannelName(id uint, name string) {
  if id==0 return // validate theres really an id
  db.Model(&Channel{}).Where("id = ?", id).Update("name", name)
}
```

Comment (Evanfeenstra) [reply routes.go:540]: Yes, channels should have unique ID (auto-incrementing INT)

```sql
CREATE TABLE people (
  id SERIAL PRIMARY KEY,
  tribe_uuid TEXT NOT NULL,
  name TEXT NOT NULL,
  created timestamptz,
);
```

File: structs.go [hunk 2]
```diff
@@ -133,6 +133,15 @@ type Pagination struct {
 	Sort  string `json:"sort"`
 }
 
+type Channel struct {
+	ID          uint       `json:"id"`
+	OwnerPubKey string     `json:"owner_pubkey"`
```
[the remark this answers is not in the document]
Comment (Evanfeenstra) [reply structs.go:138]: hmm ya i guess its fine to add the owner pub key on the channel model too

**Comment by Evanfeenstra** [comment]
@kevkevinpal lets add "channels" as an array to the `getTribe` endpoint, as that is already called when the app enters a tribe. The easy way is like this:

```go
func getTribe(w http.ResponseWriter, r *http.Request) {
	uuid := chi.URLParam(r, "uuid")
	tribe := DB.getTribe(uuid)
	
	// convert to json
	var theTribe map[string]interface{}
        j, _ := json.Marshal(tribe)
        json.Unmarshal(j, &theTribe)
        // add the channels array
        theTribe["channels"] = DB.getChannelsByTribe(uuid)

	w.WriteHeader(http.StatusOK)
	json.NewEncoder(w).Encode(tribe)
}
```

This way we dont need to deal with a new struct etc

**Comment by kevkevinpal** [comment]
> @kevkevinpal lets add "channels" as an array to the `getTribe` endpoint, as that is already called when the app enters a tribe. The easy way is like this:
> 
> ```go
> func getTribe(w http.ResponseWriter, r *http.Request) {
> 	uuid := chi.URLParam(r, "uuid")
> 	tribe := DB.getTribe(uuid)
> 	
> 	// convert to json
> 	var theTribe map[string]interface{}
>         j, _ := json.Marshal(tribe)
>         json.Unmarshal(j, &theTribe)
>         // add the channels array
>         theTribe["channels"] = DB.getChannelsByTribe(uuid)
> 
> 	w.WriteHeader(http.StatusOK)
> 	json.NewEncoder(w).Encode(theTribe)
> }
> ```
> 
> This way we dont need to deal with a new struct etc

ya that makes sense, so when I change the create channel endpoint should basically call the update tribe endpoint and change the channels field and also when we delete a channel should I leave an empty string or is another data type more preferable?

**Comment by Evanfeenstra** [comment]
> ya that makes sense, so when I change the create channel endpoint should basically call the update tribe endpoint and change the channels field and also when we delete a channel should I leave an empty string or is another data type more preferable?

Not sure I get you. You already had a route to create a channel, there is no need to update the tribe. When you delete a channel its simply deleted, so it wont show in the tribe anymore.

My code above is just to add the `channels` json into the `getTribe` route, so that the app can see all the channels in a tribe without needing to do 2 requests

**Comment by kevkevinpal** [comment]
> > ya that makes sense, so when I change the create channel endpoint should basically call the update tribe endpoint and change the channels field and also when we delete a channel should I leave an empty string or is another data type more preferable?
> 
> Not sure I get you. You already had a route to create a channel, there is no need to update the tribe. When you delete a channel its simply deleted, so it wont show in the tribe anymore.
> 
> My code above is just to add the `channels` json into the `getTribe` route, so that the app can see all the channels in a tribe without needing to do 2 requests.
> 
> No need to add a `channels` array to the Tribe stuct, that will just overcomplicate things

ok gotcha thought you were saying delete the channels struct and then put an array in the tribes struct but I get it now

**Comment by kevkevinpal** [comment]
should I add the `channels` to `getTribeByUniqueName` and `getTribesByOwner` aswell?

**Comment by kevkevinpal** [comment]
@Evanfeenstra let me know if this looks good and also are the next steps

- [ ] add parentId in relay
- [ ] update send message in relay to use parentId
- [ ] update client to use the `tribe.channels` they get from here
- [ ] update client to use parentId when sending a message

**Comment by Evanfeenstra** [comment]
> @Evanfeenstra let me know if this looks good and also are the next steps
> 
> * [ ]  add parentId in relay
> * [ ]  update send message in relay to use parentId
> * [ ]  update client to use the `tribe.channels` they get from here
> * [ ]  update client to use parentId when sending a message

changes are ready in relay https://github.com/stakwork/sphinx-relay/tree/msg-channels. So we can test in `sphinx-stack`, before starting implementing in clients

**Comment by Evanfeenstra** [comment]
> should I add the `channels` to `getTribeByUniqueName` and `getTribesByOwner` aswell?

hmm ya i guess why not. Maybe you can make a nice little function for it

**Comment by kevkevinpal** [comment]
> > @Evanfeenstra let me know if this looks good and also are the next steps
> > 
> > * [ ]   add parentId in relay
> > * [ ]   update send message in relay to use parentId
> > * [ ]   update client to use the `tribe.channels` they get from here
> > * [ ]   update client to use parentId when sending a message
> 
> changes are ready in relay https://github.com/stakwork/sphinx-relay/tree/msg-channels. So we can test in `sphinx-stack`, before starting implementing in clients

Ok cool I can write some tests for the message channels aswell to test on stack

**Review by Evanfeenstra — CHANGES_REQUESTED** [review]
@kevkevinpal You need to make sure the owner_pubkey of the created Channel matches the owner's pubkey from auth. Right now, any OwnerPubkey can be in the Channel struct when its posted. 

```
// add the pubkey here
channel.OwnerPubkey = pubkeyFromAuth
DB.createChannel(channel)
```

**Comment by kevkevinpal** [comment]
> `createChannel`

I'm not sure if I'm doing that a few lines above with this block of code

```
	tribe := DB.getTribe(channel.TribeUUID)
	if tribe.OwnerPubKey != pubKeyFromAuth {
		fmt.Println(err)
		w.WriteHeader(http.StatusNotAcceptable)
		return
	}
```

Is this the same as what you were saying, I'm checking the `OwnerPubKey` from the tribe we're trying to add the channel to and then throwing an error if it doesn't equal the `pubKeyFromAuth`

**Pull request merged**